### PR TITLE
[ADD] product_top_category

### PR DESCRIPTION
**Pull request by victor-champonnois**

### CODE REVIEW

File: product_top_category/models/product_category.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    def _get_top_category(self):
+        self.ensure_one()
+        if self.parent_id:
+            return(self.parent_id._get_top_category())
+        else: 
+            return self
+
+    # what if there is a change in parent of parent ? how to trigger the compute ?
+    # note : the field is store as related in product.template
+    @api.depends('parent_id')
```
Comment (legalsylvain) [inline product_top_category/models/product_category.py]: You could dépend on parent_id.top_categ_id. Dont you think ?

Comment (victor-champonnois) [reply product_top_category/models/product_category.py]: Yes, good idea thanks :)

**Comment by victor-champonnois** [comment]
@legalsylvain any idea why travis is failing ? The error is 
```
January 2021. pip 21.0 will remove support for this functionality.

ERROR: Could not find a version that satisfies the requirement pylint-plugin-utils==0.7 (from -r /home/travis/maintainer-quality-tools/travis/pylint_odoo_requirements.txt (line 1)) (from versions: 0.1, 0.1.1, 0.2, 0.2.1, 0.2.2, 0.2.3, 0.2.4, 0.2.6, 0.3, 0.4, 0.5, 0.6)

ERROR: No matching distribution found for pylint-plugin-utils==0.7 (from -r /home/travis/maintainer-quality-tools/travis/pylint_odoo_requirements.txt (line 1))

The command "travis_install_nightly" failed and exited with 1 during 
```

**Comment by legalsylvain** [comment]
Travis is totally obsolete. I guess that switching to github action could solve the problem, updating technical files in the root folder (in a dedicated PR).
do you know how to do that ? 

otherwise, you can ask to @carmenbianca I think.

File: product_top_category/models/product_category.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+from odoo import api, fields, models
+
+
+class ProductCategory(models.Model):
+    _inherit = "product.category"
+
+    def _get_top_category(self):
+        self.ensure_one()
+        if self.parent_id:
+            return(self.parent_id._get_top_category())
```
Comment (robinkeunen) [inline product_top_category/models/product_category.py]: Why the parentheses ?

File: product_top_category/readme/DESCRIPTION.rst [hunk 3]
```diff
@@ -0,0 +1,3 @@
+For all categories, compute their top categories.
```
Comment (robinkeunen) [inline product_top_category/readme/DESCRIPTION.rst]: Maybe add that this field is useful for reports and pivot views ?

**Review by robinkeunen — APPROVED** [review]
Small comments otherwise LGTM 👍

File: product_top_category/models/product_category.py (hunk 2 above)
Comment (victor-champonnois) [reply product_top_category/models/product_category.py]: ah, don't know, it just came naturally ^^ I remove them

File: product_top_category/readme/DESCRIPTION.rst (hunk 3 above)
Comment (victor-champonnois) [reply product_top_category/readme/DESCRIPTION.rst]: Yes, done

**Comment by victor-champonnois** [comment]
@robinkeunen thanks for the review !

**Comment by legalsylvain** [comment]
Hi. could you solve travis issue, updating files in another PR via copier template ? (then rebasing).

thanks !

**Comment by victor-champonnois** [comment]
> Hi. could you solve travis issue, updating files in another PR via copier template ? (then rebasing).
> 
> thanks !

Done [here](https://github.com/OCA/product-attribute/pull/1226), it still doesn't seem to work though. @carmenbianca do you have an idea why ?

**Comment by carmenbianca** [comment]
@victor-champonnois In the meantime, the template broke again :sweat_smile:

https://github.com/OCA/oca-addons-repo-template/pull/181

Need to bump template again

**Comment by victor-champonnois** [comment]
@carmenbianca  updated the copier update branch and rebased on it !

**Comment by rousseldenis** [comment]
@victor-champonnois Easier to do a copier update PR as this one could last again some time. The copier PR could be merged fast. Do you take care of that ?

**Comment by rousseldenis** [comment]
> @victor-champonnois Easier to do a copier update PR as this one could last again some time. The copier PR could be merged fast. Do you take care of that ?

Ok, I see it. I'll do the merge

**Comment by carmenbianca** [comment]
Rebasing this.

**Comment by victor-champonnois** [comment]
@rousseldenis this is ready to be merged

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -38,6 +38,7 @@ repos:
     rev: v3.4.1
     hooks:
       - id: flake8
+        language_version: python3
```
Comment (rousseldenis) [inline .pre-commit-config.yaml]: @victor-champonnois You should not modify manually.

Maybe could you propose this change to https://github.com/OCA/oca-addons-repo-template/blob/master/src/.pre-commit-config.yaml.jinja ?

Thanks

Comment (victor-champonnois) [reply .pre-commit-config.yaml]: I removed the commit, it wasn't needed actually.

**Comment by remytms** [comment]
@rousseldenis It seams that this PR is now ok to merge. :)

File: product_top_category/models/product_category.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+            return self.parent_id._get_top_category()
+        else:
+            return self
+
+    @api.depends('parent_id', 'parent_id.top_categ_id')
+    def _compute_top_category(self):
+        for categ in self:
+            categ.top_categ_id = categ._get_top_category()
+
+    top_categ_id = fields.Many2one('product.category',
```
Comment (rousseldenis) [inline product_top_category/models/product_category.py]: @victor-champonnois Could you put fields on ... top :-) ? Thanks

**Review by rousseldenis — APPROVED** [review]

File: product_top_category/models/product_category.py (hunk 5 above)
Comment (victor-champonnois) [reply product_top_category/models/product_category.py]: done, thanks for the review :)

**Pull request closed without merging**